src/strtree.c

```diff
@@ -168,84 +282,160 @@ static PyArrayObject *STRtree_query(STRtreeObject *self, PyObject *args) {
 
     // query the tree for indices of geometries in the tree with
     // envelopes that intersect the geometry.
-    kv_init(arr);
+    kv_init(query_indexes);
     if (geom != NULL) {
-        GEOSSTRtree_query_r(context, self->ptr, geom, query_callback, &arr);
+        GEOSSTRtree_query_r(context, self->ptr, geom, query_callback, &query_indexes);
     }
 
-    // No predicate function provided, return all geometry indexes from
-    // query.
-    // If array is empty, return an empty numpy array
-    if (predicate == 0 || kv_size(arr) == 0) {
-        result = copy_kvec_to_npy(&arr);
-        kv_destroy(arr);
+    if (predicate_id == 0 || kv_size(query_indexes) == 0) {
+        // No predicate function provided, return all geometry indexes from
+        // query.  If array is empty, return an empty numpy array
+        result = copy_kvec_to_npy(&query_indexes);
+        kv_destroy(query_indexes);
         return (PyArrayObject *) result;
     }
 
-    switch (predicate) {
-        case 1: {  // intersects
-            predicate_func = (FuncGEOS_YpY_b *)GEOSPreparedIntersects_r;
-            break;
-        }
-        case 2: { // within
-            predicate_func = (FuncGEOS_YpY_b *)GEOSPreparedWithin_r;
-            break;
-        }
-        case 3: { // contains
-            predicate_func = (FuncGEOS_YpY_b *)GEOSPreparedContains_r;
-            break;
-        }
-        case 4: { // overlaps
-            predicate_func = (FuncGEOS_YpY_b *)GEOSPreparedOverlaps_r;
-            break;
-        }
-        case 5: { // crosses
-            predicate_func = (FuncGEOS_YpY_b *)GEOSPreparedCrosses_r;
-            break;
-        }
-        case 6: { // touches
-            predicate_func = (FuncGEOS_YpY_b *)GEOSPreparedTouches_r;
-            break;
-        }
-        default: { // unknown predicate
-            PyErr_SetString(PyExc_ValueError, "Invalid query predicate");
-            return NULL;
-        }
+    predicate_func = get_predicate_func(predicate_id);
+    if (predicate_func == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "Invalid predicate for query");
```

Comment:
Doesn't the get_predicate_func already set the error? (not very familiar to how this works though)

Reply:
And if it is needed here, it is rather a ValueError than RuntimeError?

Reply:
Correct, it is duplicate.  Removed.